aws/rust-runtime/aws-config/src/meta/region.rs

```diff
@@ -0,0 +1,159 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0.
+ */
+
+use aws_types::region::Region;
+use std::borrow::Cow;
+use std::fmt::Debug;
+use tracing::Instrument;
+
+/// Load a region by selecting the first from a series of region providers
+#[derive(Debug)]
+pub struct RegionProviderChain {
+    providers: Vec<Box<dyn ProvideRegion>>,
+}
+
+impl RegionProviderChain {
+    /// Load a region from the provider chain
+    ///
+    /// The first provider to return a non-optional region will be selected
+    pub async fn region(&self) -> Option<Region> {
+        for provider in &self.providers {
+            if let Some(region) = provider
+                .region()
+                .instrument(tracing::info_span!("load_region", provider = ?provider))
+                .await
+            {
+                return Some(region);
+            }
+        }
+        None
+    }
+}
+
+/// Implement a region provider based on a series of region providers
+///
+/// # Example
+/// ```rust
+/// use aws_types::region::Region;
+/// use std::env;
+/// use aws_config::meta::region::RegionProviderChain;
+/// // region provider that first checks the `CUSTOM_REGION` environment variable,
+/// // then checks the default provider chain, then falls back to us-east-2
+/// let provider = RegionProviderChain::first_try(env::var("CUSTOM_REGION").ok().map(Region::new))
+///     .or_default_provider()
+///     .or_else(Region::new("us-east-2"));
+/// ```
+impl RegionProviderChain {
+    /// Create a default provider chain that starts by checking this provider.
+    pub fn first_try(provider: impl ProvideRegion + 'static) -> Self {
+        RegionProviderChain {
+            providers: vec![Box::new(provider)],
+        }
+    }
+
+    /// Add a fallback provider to the region provider chain.
+    pub fn or_else(mut self, fallback: impl ProvideRegion + 'static) -> Self {
+        self.providers.push(Box::new(fallback));
+        self
+    }
+
+    /// Create a region provider chain that starts by checking the default provider.
+    #[cfg(feature = "default-provider")]
+    pub fn default_provider() -> Self {
+        Self::first_try(crate::default_provider::region::default_provider())
+    }
+
+    /// Fallback to the default provider
+    #[cfg(feature = "default-provider")]
+    pub fn or_default_provider(mut self) -> Self {
+        self.providers
+            .push(Box::new(crate::default_provider::region::default_provider()));
+        self
+    }
+}
+
+impl ProvideRegion for Option<Region> {
+    fn region(&self) -> future::ProvideRegion {
+        future::ProvideRegion::ready(self.clone())
+    }
+}
+
+impl ProvideRegion for RegionProviderChain {
+    fn region(&self) -> future::ProvideRegion {
+        future::ProvideRegion::new(self.region())
```

Comment:
This looks like it could be an infinite loop. Could we make the call to `region` more explicit?